### PR TITLE
qs web-components

### PR DESCRIPTION
**Pull request by ia3andy**

### CODE REVIEW

File: src/main/java/io/quarkus/search/app/hibernate/InputProviderHtmlBodyTextBridge.java [hunk 1]
```diff
@@ -18,7 +18,7 @@ public String toIndexedValue(InputProvider provider, ValueBridgeToIndexedValueCo
         try (var in = provider.open()) {
             Element body = Jsoup.parse(in, StandardCharsets.UTF_8.name(), "/").body();
             // Content div has two grid columns: actual content and TOC. There's not much use of the TOC, we want the content only:
-            Element content = body.selectFirst(".content .grid__item");
+            Element content = body.selectFirst(".content .grid_item");
```
Comment (yrodiere) [inline src/main/java/io/quarkus/search/app/hibernate/InputProviderHtmlBodyTextBridge.java:21]: I don't think that's right? At least it causes many problems when indexing:

```
2023-11-22 09:02:12,305 WARN  [io.qua.sea.app.hib.InputProviderHtmlBodyTextBridge] (executor-thread-1) Was unable to find the content section of a guide. Using whole document as text. GitInputProvider{path='version/2.13/guides/all-builditems.html'}
```

If I revert this change, the warnings disappear.

File: .gitignore [hunk 2]
```diff
@@ -5,6 +5,7 @@ pom.xml.releaseBackup
 pom.xml.versionsBackup
 release.properties
 .flattened-pom.xml
+node_modules/
```
Comment (yrodiere) [inline .gitignore:8]: Maybe a comment to explain the hack?

File: src/main/resources/web/app/qs-form.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        } else {
+          this._total = r.hits.length;
+        }
+        const hasMoreHits = r.total > this._total;
+
+        this.dispatchEvent(new CustomEvent(event, {detail: {...r, hasMoreHits}}));
+      }).finally(() => {
+      this._loading = false;
+      this.dispatchEvent(new CustomEvent(QS_END_EVENT));
+    });
```
Comment (yrodiere) [inline src/main/resources/web/app/qs-form.ts:116]: I don't see offline (javascript) search handling in case we reach the timeout... Is it supposed to be implemented elsewhere?

**Review by yrodiere — COMMENTED** [review]
Thanks. And thanks for trying to keep some of my code, that's kind of you.

I added a few comments, hopefully some of them are useful :)

File: src/main/java/io/quarkus/search/app/dto/SearchHit.java [hunk 4]
```diff
@@ -5,5 +5,6 @@
 import org.hibernate.search.mapper.pojo.mapping.definition.annotation.ProjectionConstructor;
 
 @ProjectionConstructor
-public record SearchHit(@IdProjection String id, @FieldProjection String title) {
+public record SearchHit(@IdProjection String id, @FieldProjection String title, @FieldProjection String summary,
+        @FieldProjection String keywords) {
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/java/io/quarkus/search/app/dto/SearchHit.java:9]: Ok that was just an attempt, I will remove it.

File: src/main/resources/web/app/qs-form.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    if (this._isInput(target)) {
+      if (target.value.length === 0) {
+        this._clearSearch();
+        return;
+      }
+      if (target.value.length < this.minChars) {
+        return;
+      }
+    }
+    this._searchDebounced();
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-form.ts:133]: Yeah I need to have a look at the whole abort to implement it, I've ignored this and the timeout for now

File: src/main/resources/web/app/qs-target.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    super.disconnectedCallback();
+  }
+
+  render() {
+    if (this._result?.hits) {
+      if (this._result.hits.length === 0) {
+        return html`
+          <div id="qs-target">
+            <p>No results found</p>
+          </div>
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-target.ts:46]: For sure! same, I took shortcut to have something to share and iterate

File: src/main/java/io/quarkus/search/app/hibernate/InputProviderHtmlBodyTextBridge.java (hunk 1 above)
Comment (ia3andy) [reply src/main/java/io/quarkus/search/app/hibernate/InputProviderHtmlBodyTextBridge.java:21]: ah I think this is IntelliJ refactor good catch!!!!!!

File: .gitignore (hunk 2 above)
Comment (ia3andy) [reply .gitignore:8]: ok

File: src/main/resources/web/app/qs-form.ts (hunk 3 above)
Comment (ia3andy) [reply src/main/resources/web/app/qs-form.ts:116]: Yes, I need to see if we pass the data with json or if we need to parse the page content and send an event for this.

File: src/main/resources/web/app/qs-target.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-target.ts]: Fixed, need a bit more styling when integrating probably

**Review by yrodiere — COMMENTED** [review]
Thanks, looking great.

A few comments below :)

File: src/main/resources/web/app/assets/icons/docsicon-concepts.svg
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/assets/icons/docsicon-concepts.svg:1]: They are configured to be bundled as base64 url, they won't be available directly.

File: src/main/java/io/quarkus/search/app/WebComponents.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    @Inject
+    Bundle bundle;
+
+    @Context
+    UriInfo uriInfo;
+
+    // This route allows to access the web-component js on a fixed path
+    // without affecting caching of the script
+    @Path("/search.js")
+    @Cache(noCache = true)
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/java/io/quarkus/search/app/WebComponents.java:28]: As discussed on Zulip, only the redirect will not be cached (the script will be cached), so that if the script changes, the the redirect will change (and be called again).

Also as discussed there is a price for this but I am not sure there is a way to know if it has changed without any call?

File: src/main/resources/web/app/qs-guide.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  private _renderContent(content: string) {
+    if(!content) {
+      return content;
+    }
+    if(Array.isArray(content)) {
+      return content.map((c) => this._renderContent(c));
+    }
+    const parts = content.split(/___([^_]+)___/g);
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-guide.ts:114]: I think it's pretty unlikely, but could we maybe have a detector before hand for it? 

It's nice to have the content human readable. I could settle on something more readable :-)

`_H__myword__L_` this one is very very unlikely even more with opening and closing tags

File: src/main/resources/web/app/qs-guide.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  render() {
+    return html`
+      <div class="qs-hit qs-guide type-${this.data.type}">
+        <h4><a href="${this.data.url}">${this._renderContent(this.data.title)}</a></h4>
+        <div class="summary">
+          <p>${this._renderContent(this.data.summary)}</p>
+        </div>
+        <div class="keywords">${this._renderContent(this.data.keywords)}</div>
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-guide.ts:102]: Ok I will add it!

File: src/main/resources/web/app/qs-guide.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      </div>
+    `;
+  }
+
+  private _renderContent(content: string) {
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-guide.ts:107]: I'll give it a try

**Comment by ia3andy** [comment]
Thanks @yrodiere for the review!!!

File: src/main/resources/web/app/qs-guide.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (yrodiere) [reply src/main/resources/web/app/qs-guide.ts:114]: > could we maybe have a detector before hand for it?

Not on the server side. Like I said the other day, there's HTML escaping, or nothing.

> It's nice to have the content human readable. I could settle on something more readable :-)

Yeah that's why we were using HTML ;D

> `_H__myword__L_` this one is very very unlikely even more with opening and closing tags

Ok, that seems safer.

File: src/main/resources/web/app/assets/icons/docsicon-concepts.svg
[the remark this answers is not in the document]
Comment (yrodiere) [reply src/main/resources/web/app/assets/icons/docsicon-concepts.svg:1]: We still need copyright mentions for assets we include in our source... All the more so if we distribute them at runtime (however indirectly).

File: src/main/java/io/quarkus/search/app/WebComponents.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (yrodiere) [reply src/main/java/io/quarkus/search/app/WebComponents.java:28]: > As discussed on Zulip, only the redirect will not be cached (the script will be cached), so that if the script changes, the the redirect will change (and be called again).

Seems fair.

> Also as discussed there is a price (on REST call) for this but I am not sure there is a way to know if it has changed without any call?

Makes sense.

**Comment by ia3andy** [comment]
Ok, I've pushed again:
- switched to html (and reverted the service changes)
- added a loading indicator
- added the content section

For the icons copyright, I've used the one from quarkusio so we should mimic what they are doing?

**Comment by yrodiere** [comment]
> For the icons copyright, I've used the one from quarkusio so we should mimic what they are doing?

Yep, makes sense

**Comment by ia3andy** [comment]
> Hey @ia3andy I see you marked this as ready for review, but just to make sure we agree: you intend to make some more changes, right? Regarding the copyright mentions in particular, and maybe the abort thing (see earlier reviews).
> 
> I do understand javascript search will be handled elsewhere.

Yes sorry I am stuck on code.quarkus, I will get back to this very soon

**Comment by yrodiere** [comment]
> > Hey @ia3andy I see you marked this as ready for review, but just to make sure we agree: you intend to make some more changes, right? Regarding the copyright mentions in particular, and maybe the abort thing (see earlier reviews).
> > I do understand javascript search will be handled elsewhere.
> 
> Yes sorry I am stuck on code.quarkus, I will get back to this very soon

No problem, I just wanted to be sure there wasn't a misunderstanding. Take your time :+1:

**Comment by yrodiere** [comment]
Hey @ia3andy , since you seem busy, and there are still a few things to fix here, and I think there's still a bit of work to get this integrated with quarkus.io, I'll plan on deploying the first version without web-components. We can always add them later when this is ready, of course.

File: pom.xml [hunk 11]
```diff
@@ -335,4 +367,67 @@
       </plugin>
     </plugins>
   </build>
+  <profiles>
+    <profile>
+      <id>locker</id>
```
Comment (marko-bekhta) [inline pom.xml:372]: It would be nice to have a comment on what this profile is for and when to use it, if ever 😃

File: pom.xml [hunk 12]
```diff
@@ -107,6 +107,31 @@
...
+      <version>4.17.21</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mvnpm.at.types</groupId>
+      <artifactId>lodash</artifactId>
+      <version>4.14.202</version>
+      <scope>provided</scope>
+      <classifier>sources</classifier>
+    </dependency>
```
Comment (marko-bekhta) [inline pom.xml:134]: for other dependencies, we've tried to keep the versions under the dependency management section, and here, just use them without versions.

File: src/main/resources/web/app/qs-guide.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    .qs-guide div {
+      margin: 1rem 0 0 90px;
+      font-size: 1rem;
+      line-height: 1.5rem;
+      font-weight: 400;
+    }
+
+    .qs-guide .content-highlights {
+      font-size: 0.7rem;
+    }
```
Comment (marko-bekhta) [inline src/main/resources/web/app/qs-guide.ts:84]: For this section where we are displaying the "full-content" highlights, we've had it stand out less so that the title and the summary are the key areas to look for highlighted terms.
So maybe let's do something like: 

```suggestion
    .qs-guide .content-highlights {
      font-size: 0.7rem;
      line-height: 1rem;
      color: var(--some-gray-color-variable-name);
    }
```

**Review by marko-bekhta — COMMENTED** [review]
Looks nice! 😃

**Review by yrodiere — COMMENTED** [review]
Thanks, looks nice indeed. I added a few comments below.

File: src/main/resources/web/index.html [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    {#bundle/}
+</head>
+<body>
+<div class="content">
+    <nav class="search">
+        <div class="logo"></div>
+        <qs-form>
+            <div class="search-query">
+                <input name="q" type="search" placeholder="Start searching..."/>
+            </div>
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply src/main/resources/web/index.html:23]: Oh, and a language too.

[the remark this answers is not in the document]
Comment (yrodiere) [reply src/main/resources/web/index.html:23]: See https://github.com/quarkusio/search.quarkus.io/pull/210

File: src/test/java/io/quarkus/search/app/WebComponentsTest.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import com.microsoft.playwright.BrowserContext;
+import com.microsoft.playwright.Page;
+import com.microsoft.playwright.Response;
+
+import io.quarkiverse.playwright.InjectPlaywright;
+import io.quarkiverse.playwright.WithPlaywright;
+
+@QuarkusTest
+@WithPlaywright
+public class WebComponentsTest {
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/test/java/io/quarkus/search/app/WebComponentsTest.java:20]: Yes I need to add them

File: src/test/java/io/quarkus/search/app/WebComponentsTest.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.microsoft.playwright.BrowserContext;
+import com.microsoft.playwright.Page;
+import com.microsoft.playwright.Response;
+
+import io.quarkiverse.playwright.InjectPlaywright;
+import io.quarkiverse.playwright.WithPlaywright;
+
+@QuarkusTest
+@WithPlaywright
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/test/java/io/quarkus/search/app/WebComponentsTest.java:19]: I need to check if we can only use Chrome

[the remark this answers is not in the document]
Comment (ia3andy) [reply src/test/java/io/quarkus/search/app/WebComponentsTest.java:19]: It seems it's not yet possible:
https://github.com/microsoft/playwright-java/issues/1196

The weird thing is that it should use the maven dependency and not download it AFAIK

File: src/main/resources/web/app/qs-form.ts [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            this._loading.abort();
+        }
+        const controller = new AbortController();
+        this._loading = controller;
+        this.dispatchEvent(new CustomEvent(QS_START_EVENT));
+        const data = this._readQueryInputs();
+        if (this.localSearch) {
+            this._localSearch();
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply src/main/resources/web/app/qs-form.ts]: I think the latest impl is working fine, let me know if you find any issue. I've tried to do something close to what you did but with the events.

**Comment by ia3andy** [comment]
All good I think

**Comment by ia3andy** [comment]
I think we should merge this, and iterate..

**Comment by ia3andy** [comment]
Yeaaahh!! 🥳

**Comment by ia3andy** [comment]
Now let me complete the quarkusio website!

**Comment by yrodiere** [comment]
Thanks @ia3andy :)
